### DISCUSSION TITLE
TVHeadEnd - quase todos os muxes falham

### DISCUSSION THREAD

**Discussion by jssilva25** [LITUATUI/M3UPT#34] [General]
Olá,
Obrigado pelo bom trabalho.
Eu tenho um servidor de TV com TVHeadEnd num Rpi com Debian, que debita as gravações e canais para todos os Kodi's da casa. Tem dois tuners, um DVB-T para a TV aérea e outro DVB-C para o cabo da ZON; as gravações são gravadas por SMB para o servidor TrueNas da casa; funciona.
Mas tentei usar a lista mu3 que publicam com fracos resultados em TV: só funcionam a SIC, Yas TV, Trace Urban, Bloomberg TV, CGTN Français, isto em TV's.
Testei alguns rádios, beachcam's e TvShows sem encontrar falhas.
No entanto, o que mais me interessava, as RTP's, não funcionam.
Também gostava de vos perguntar se será possível fazer uma lista m3u para o nostv.pt
Obrigado.

**Comment by LITUATUI** [comment]
Olá.

Basta no Kodi usar o [PVR IPTV Simple Client](https://kodi.wiki/view/Add-on:PVR_IPTV_Simple_Client) para introduzir a lista.

Usar uma lista de IPTV no TVHeadend é má ideia.

No meu Kodi tenho os dois clientes PVR activos (PVR IPTV Simple Client e Tvheadend HTSP Client). Funcionam bem em simultâneo.

**Reply by jssilva25** [reply]
Olá, obrigado pela rápida resposta.

Há vários problemas em usar o Kodi directamente para ver IPTV:

1. Quando é necessária alguma alteração, é necessário fazê-la em todas as máquinas, uma a uma; e já são algumas.
2. É necessário configurar os canais em todas as máquinas, uma a uma; sobretudo porque, como já disse, tenho muitos mais canais provenientes dos tuners do TvHeadEnd.
3. Pode programar-se uma gravação específica numa máquina com Kodi, é verdade; eu tenho todos os telejornais a gravar todos os dias (gosto de poder passar os anúncios e o que mais não me interessa). Mas é necessário garantir que essa máquina está ligada à hora da gravação e enviá-la para o servidor para acesso das outras. Pôr todas a gravar o mesmo não será muito eficiente...
4. A gravação pode ser enviada para o servidor para que todas as outras máquinas acedam, mas não é fiável.

Por tudo isto, prefiro usar o Kodi  em cada máquina somente como cliente e ter um servidor. Assim, qualquer alteração é feita somente no servidor, que providencia os canais e faz as gravações para acesso de todas as outras.

De facto, o Kodi foi concebido para ser autónomo e auto-suficiente, não para funcionar numa rede de media partilhada. Para isso temos Plex, Emby, Jellyfin e outros, que afinal podem ter Kodi's como clientes.

Se não é possível, paciência... Continuo como estou.

E a propósito de uma lista para nostv.pt, obviamente com autenticação que tenho, será possível?

Obrigado.

**Reply by jssilva25** [reply]
A propósito, desenvolvi em tempos um script para a RTP1 da rtp play a colocar na janela de URL do TVHeadEnd, que funcionou na altura. Entretanto perdi o interesse e agora que experimentei de novo, já não funciona:

pipe:///usr/bin/avconv -re -v error -fflags +genpts -rtmp_playpath 2ch5h264 -rtmp_swfverify http://programas.rtp.pt/play/player.swf?v3 -rtmp_pageurl http://www.rtp.pt/play/direto/rtp1?mag -rtmp_live live -i rtmp://rtppullswflivefs.fplive.net/rtppullswflive-live -vcodec copy -acodec copy -bsf h264_mp4toannexb -f mpegts pipe:1

Também se pode usar ffmpeg em vez de avconv.

Oferece-lhe algum comentário?

**Reply by LITUATUI** [reply]
Os canais da RTP precisam de um user-agent, não sei se é possível definir isto no TVHeadend. 

Por exemplo no MPV:

`mpv https://streaming-live-app.rtp.pt/liverepeater/smil:rtpnHD.smil/rtp_playlist.m3u8 --http-header-fields='User-Agent: Firefox'`

Desconheço como funciona o serviço da nostv.pt mas duvido que seja possível.

Quanto ao script é normal que não funcione, pois a RTP já modificou os seus streams várias vezes nos últimos anos.

**Reply by jssilva25** [reply]
Obrigado pelo script, funciona e é para mim uma base de trabalho.

Parece que o TVHeadEnd não gosta de streams HLS. Vou tentar fazer pipe, semelhante ao script que já enviei acima, para ffmpeg.

Se conseguir, posto aqui para uso de outros interessados.

Mais uma vez, obrigado pelo apoio.

**Reply by jssilva25** [reply]
Já está, trabalha. No fundo é só transformar os streams hls em ts para o TVHeadEnd aceitar.

Primeiro a prova de conceito do ffmpeg piped para o mpv:
`ffmpeg -user-agent Firefox -i https://streaming-live-app.rtp.pt/liverepeater/smil:rtp1HD.smil/rtp_playlist.m3u8 -c copy -f mpegts - | mpv -`

Feito o teste, agora, no campo _URL_ do max do TVHeadEnd:
`pipe:///usr/local/bin/ffmpeg -user-agent Firefox -i https://streaming-live-app.rtp.pt/liverepeater/smil:rtp1HD.smil/rtp_playlist.m3u8 -c copy -f mpegts pipe:1`

Claro que pode ter-se que adaptar a localização do binário ffmpeg conforme o OS e Dist.
Pode também pôr-se o ícone no campo _URL do _ícone__

Quanto ao EPG, eu uso WG++, muito fiável e bom.

Depois disto, é só alterar a vossa lista original aplicando este conceito, o que vou tratar de seguida.

**Reply by LITUATUI** [reply]
Algum motivo para usar Debian em vez do LibreELEC?

**Reply by jssilva25** [reply]
Também tenho LibreElec nos Rpi das TVs para correr exclusivamente Kodi que são alimentados do meu servidor Jellyfin residente numa jail do servidor da casa em TrueNAS (freeBSD). Mas não correm mais nada e assim consigo fazer as alterações às bibliotecas centralmente, uma só vez. Toda a multimedia está obviamente no TrueNas, gerida pelo Jellyfin.

Depois tenho um Rpi com Debian a correr somente TVHeadEnd que debita canais e gravações para os Kodi's por htsp, embora as gravações estejam no TrueNAS por SMB.

Sim, é muito mais flexível, permite instalação de todas as aplicações Linux onde o LibreElec é muito "perro", é necessário dar muitas voltas se for de todo possível.

Parece um sistemas complexo mas não é; e é muito fiável com backups por todos os lados para discos zfs em Raid. Já por várias vezes tive soluços numa das componentes (tvheadend, etc) e nem me preocupo averiguar o que aconteceu; faço um clone do snapshot de ontem, ou da semana passada, e já está.

Mas esta questão foi inicialmente levantada porque muitas vezes tenho problemas de seek, de som, de qualidade, e às vezes nem gravam, nas gravações dos noticiários e portanto resolvi implementar as gravações por IPTV num outro TvHeadEnd a correr noutra jail do TrueNAS. Tem potência de CPU, memória e discos de sobra, e assim posso parar o Rpi (sempre são menos 7 Watt / 24 horas, uns € ao final do mês).

Enfim, é uma tarefa para aplicar eficientemente e com prazer o pouco tempo que me sobra.

**Reply by jssilva25** [reply]
Obrigado!

**Reply by LITUATUI** [reply]
Quanto ao EPG, antes usava o WebGrab+Plus numa VPS Debian, mas depois a versão gratuita ficou bastante limitada, para além de lenta.

Agora uso o [XMLTV](https://github.com/XMLTV/xmltv) que para além de ser de código aberto, é muito superior e dá para correr no GitHub Actions.

**Reply by jssilva25** [reply]
E acrescento que o WG++ em freeBSD obriga a instalação do Mono e é de difícil instalação. Também estou neste momento às voltas com o XMLTV.

**Reply by LITUATUI** [reply]
Exacto, é igual no Debian.

O EPG da Vodafone funciona muito bem, dá uma vista de olhos no script e no ficheiro de configuração [aqui](https://github.com/LITUATUI/M3UPT/tree/main/epg).

**Comment by jssilva25** [comment]
Boa! Não tinha visto, grande ajuda. Obrigado, vou explorar.

A minha source é NOS mas os canais principais são os mesmos.
